File: scripts/day1-part1-optimized.js
const fs = require("fs");

// Optimized Part 1: Count times dial points at 0 after rotations
function solvePart1Optimized(rotations) {
  let position = 50;
  let count = 0;

  for (let i = 0; i < rotations.length; i++) {
    const rotation = rotations[i];
    const direction = rotation.charCodeAt(0); // L=76, R=82
    const distance = parseInt(rotation.substring(1), 10);

    if (direction === 76) {
      // 'L'
      position = (position - (distance % 100) + 100) % 100;
    } else {
      // 'R'
      position = (position + distance) % 100;
    }

    if (position === 0) count++;
  }

  return count;
}

// Read and parse input
const input = fs
  .readFileSync("../inputs/day-1-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

// Benchmark
const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart1Optimized(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Node.js)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);

File: scripts/day1-part1-optimized.py
import time

def solve_part1_optimized(rotations):
    """Optimized Part 1: Count times dial points at 0 after rotations"""
    position = 50
    count = 0
    
    for rotation in rotations:
        direction = rotation[0]
        distance = int(rotation[1:])
        
        if direction == 'L':
            position = (position - distance % 100 + 100) % 100
        else:  # 'R'
            position = (position + distance) % 100
        
        if position == 0:
            count += 1
    
    return count

# Read and parse input
with open("../inputs/day-1-input.txt", "r") as f:
    input_data = [line.strip() for line in f if line.strip()]

# Benchmark
iterations = 1
start_time = time.perf_counter()

result = 0
for _ in range(iterations):
    result = solve_part1_optimized(input_data)

end_time = time.perf_counter()
total_time = (end_time - start_time)
avg_micros = (total_time / iterations) * 1_000_000

print(f"Python")
print(f"Answer: {result}")
print(f"Average time: {avg_micros:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
